protocol/common/raw_messages_persistence.go

```diff
@@ -0,0 +1,262 @@
+		   skip_group_message_wrap,
+		   send_on_personal_topic,
+		   payload
+		)
+		VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)`,
+		message.ID,
+		message.LocalChatID,
+		message.LastSent,
+		message.SendCount,
+		message.Sent,
+		message.MessageType,
+		message.ResendAutomatically,
+		encodedRecipients.Bytes(),
+		message.SkipEncryption,
+		message.SendPushNotification,
+		message.SkipGroupMessageWrap,
+		message.SendOnPersonalTopic,
+		message.Payload)
+	return err
+}
+
+func (db RawMessagesPersistence) RawMessageByID(id string) (*RawMessage, error) {
+	var rawPubKeys [][]byte
+	var encodedRecipients []byte
+	var skipGroupMessageWrap sql.NullBool
+	var sendOnPersonalTopic sql.NullBool
+	message := &RawMessage{}
+
+	err := db.db.QueryRow(`
+			SELECT
+			  id,
+			  local_chat_id,
+			  last_sent,
+			  send_count,
+			  sent,
+			  message_type,
+			  resend_automatically,
+			  recipients,
+			  skip_encryption,
+		          send_push_notification,
+			  skip_group_message_wrap,
+			  send_on_personal_topic,
+		          payload
+			FROM
+				raw_messages
+			WHERE
+				id = ?`,
+		id,
+	).Scan(
+		&message.ID,
+		&message.LocalChatID,
+		&message.LastSent,
+		&message.SendCount,
+		&message.Sent,
+		&message.MessageType,
+		&message.ResendAutomatically,
+		&encodedRecipients,
+		&message.SkipEncryption,
+		&message.SendPushNotification,
+		&skipGroupMessageWrap,
+		&sendOnPersonalTopic,
+		&message.Payload,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	if rawPubKeys != nil {
+		// Restore recipients
+		decoder := gob.NewDecoder(bytes.NewBuffer(encodedRecipients))
+		err = decoder.Decode(&rawPubKeys)
+		if err != nil {
+			return nil, err
+		}
+		for _, pkBytes := range rawPubKeys {
+			pubkey, err := crypto.UnmarshalPubkey(pkBytes)
+			if err != nil {
+				return nil, err
+			}
+			message.Recipients = append(message.Recipients, pubkey)
+		}
+	}
+
+	if skipGroupMessageWrap.Valid {
+		message.SkipGroupMessageWrap = skipGroupMessageWrap.Bool
+	}
+
+	if sendOnPersonalTopic.Valid {
+		message.SendOnPersonalTopic = sendOnPersonalTopic.Bool
+	}
+
+	return message, nil
+}
+
+func (db RawMessagesPersistence) RawMessagesIDsByType(t protobuf.ApplicationMetadataMessage_Type) ([]string, error) {
+	ids := []string{}
+
+	rows, err := db.db.Query(`
+			SELECT
+			  id
+			FROM
+				raw_messages
+			WHERE
+			message_type = ?`,
+		t)
+	if err != nil {
+		return ids, err
+	}
+	defer rows.Close()
+
+	for rows.Next() {
+		var id string
+		if err := rows.Scan(&id); err != nil {
+			return ids, err
+		}
+		ids = append(ids, id)
+	}
+
+	return ids, nil
+}
+
+// MarkAsConfirmed marks all the messages with dataSyncID as confirmed and returns
+// the messageIDs that can be considered confirmed.
+// If atLeastOne is set it will return messageid if at least once of the messages
+// sent has been confirmed
+func (db RawMessagesPersistence) MarkAsConfirmed(dataSyncID []byte, atLeastOne bool) (messageIDs []types.HexBytes, err error) {
```

Comment:
@cammellos I'm wondering if it is possible to get more than one messageID as a result of this function?

My understanding so far was that we can have same messageID with different related datasyncIDs (private group case), but we can't have different messageIDs related to the same datasyncID. And that means we won't get more than one message confirmed after receiving ack-ed dataSyncID. So logic of this function can be simpler.

Please, correct me if I'm wrong

Reply:
Oh you are right,
I got confused by the fact that messages can be batched in a single datasync message, but `AppendMessage` returns the ID of the "inner" datasync message.
We can probably simplify, well spotted!